File: LeetCode/Medium/1302_Deepest_Leaves_Sum/Solution.py
from typing import *

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

## 2022-05-14
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        def traverseAndSumTree(node, depth, deepest_sum):
            if not node:
                return
            
            if depth > deepest_sum[0]:
                deepest_sum[0] = depth
                deepest_sum[1] = node.val
            elif depth == deepest_sum[0]:
                deepest_sum[1] += node.val
            
            traverseAndSumTree(node.left, depth+1, deepest_sum)
            traverseAndSumTree(node.right, depth+1, deepest_sum)
            return deepest_sum[1]
            
        return traverseAndSumTree(root, 0, [0, 0])

File: LeetCode/Medium/1302_Deepest_Leaves_Sum/Solution.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// 2022-05-13
class Solution {
    public int deepestLeavesSum(TreeNode root) {
        int[] deepest = new int[2];
        addLeaves(root, 0, deepest);
        return deepest[1];
        
    }
    
    private void addLeaves(TreeNode root, int depth, int[] deepest) {
        if (root == null) return;
        
        if (depth > deepest[0]) {
            deepest[0] = depth;
            deepest[1] = 0;
        }
        
        if (depth == deepest[0]) 
            deepest[1] += root.val;
        
        addLeaves(root.left, depth+1, deepest);
        addLeaves(root.right, depth+1, deepest);
    }
}